drivers/regmap/regmap.c

```diff
@@ -0,0 +1,379 @@
+{
+  size_t val_bytes = map->val_bytes;
+  unsigned int ival;
+  FAR uint8_t *ptr;
+  int ret = 0;
+  int i;
+
+  DEBUGASSERT(REGMAP_ALIGNED(reg, map->reg_stride));
+
+  map->lock(map);
+  if (map->write != NULL)
+    {
+      ret = map->write(map->bus, val, val_bytes * val_count);
+      goto out;
+    }
+
+  for (i = 0; i < val_count; i++)
+    {
+      ptr = (FAR uint8_t *)val + (i * val_bytes);
+      switch (val_bytes)
+        {
+          case 1:
+            ival = *(FAR uint8_t *)ptr;
+            break;
+          case 2:
+            ival = *(FAR uint16_t *)ptr;
+            break;
+          case 4:
+            ival = *(FAR uint32_t *)ptr;
+            break;
+          default:
+            goto out;
+        }
+
+      ret = map->reg_write(map->bus, reg + (i * map->reg_stride), ival);
+      if (ret < 0)
+        {
+          break;
+        }
+    }
+
+out:
+  map->unlock(map);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: regmap_read
+ *
+ * Description:
+ *   Regmap read, called after initializing the regmap bus device.
+ *   the first parameter is regmap pointer.
+ *
+ * Input Parameters:
+ *   map - regmap handler, from regmap bus init function return.
+ *   reg - register address to be read.
+ *   val - read data buffer.
+ *
+ * Returned Value:
+ *   Description of the value returned by this function (if any),
+ *   including an enumeration of all possible error values.
+ *
+ * Assumptions/Limitations:
+ *   None.
+ *
+ ****************************************************************************/
+
+int regmap_read(FAR struct regmap_s *map, unsigned int reg, FAR void *val)
+{
+  int ret;
+
+  DEBUGASSERT(REGMAP_ALIGNED(reg, map->reg_stride));
+
+  map->lock(map);
+
+  ret = map->reg_read(map->bus, reg, val);
+
+  map->unlock(map);
+  return ret;
+}
+
+/****************************************************************************
+ * Name: regmap_bulk_read
+ *
+ * Description:
+ *   Regmap bulk read, called after initializing the regmap bus device.
+ *   the first parameter is regmap pointer.
+ *
+ * Input Parameters:
+ *   map       - regmap handler, from regmap bus init function return.
+ *   reg       - register address to be read.
+ *   val       - read data buffer.
+ *   val_count - read data buffer size.
+ *
+ * Returned Value:
+ *   Description of the value returned by this function (if any),
+ *   including an enumeration of all possible error values.
+ *
+ * Assumptions/Limitations:
+ *   None.
+ *
+ ****************************************************************************/
+
+int regmap_bulk_read(FAR struct regmap_s *map, unsigned int reg,
+                     FAR void *val, unsigned int val_count)
+{
+  FAR uint32_t *u32 = val;
+  FAR uint16_t *u16 = val;
+  FAR uint8_t  *u8  = val;
+  unsigned int ival;
+  int ret = 0;
+  int i;
+
+  DEBUGASSERT(REGMAP_ALIGNED(reg, map->reg_stride));
+
+  if (map->read != NULL)
+    {
+      map->lock(map);
```

Comment:
move map->lock/unlock out of if/else block

Reply:
Done.